README.md
Xoken Arch Gateway client

> Javascript SDK to talk to Xoken Arch Gateway

[![Codecov branch](https://img.shields.io/codecov/c/gitlab/ledgit/xoken-arch-gateway-client/master)](https://codecov.io/gl/ledgit/xoken-arch-gateway-client)
[![Gitlab pipeline status (branch)](https://img.shields.io/gitlab/pipeline/ledgit/xoken-arch-gateway-client/master)](https://gitlab.com/ledgit/xoken-arch-gateway-client/pipelines)
[![Libraries.io dependency status for latest release](https://img.shields.io/librariesio/release/npm/xoken-arch-gateway-client)](https://libraries.io/npm/xoken-arch-gateway-client)
[![npm](https://img.shields.io/npm/v/xoken-arch-gateway-client)](https://www.npmjs.com/package/xoken-arch-gateway-client)


# Install

```
npm install xoken-arch-gateway-client
```

# Usage

```
import { Client } from 'xoken-arch-gateway-client'

const client = Client('localhost', 9090);

await client.getBlockByHeight(100);
await client.getBlocksByHeights([100, 101, 102]);
await client.getBlockByHash('00000000d1145790a8694403d4063f323d499e655c83426834d4ce2f8dd4a2ee');
await client.getBlocksByHashes([
      '000000000000000002af2a6de04d4a1a73973827eae348fe4d3f4d05610ff968',
      '000000000000000007fc734cbf1fc04c59cf7ecb6af0707fd5cf5b8d46dc4c75'
    ]);

await client.getTransactionByTxId('7d3eb236b526bd681b7fc499d657d237b4d3bc21ef25b37fc1c70822849f1243');
await client.getTransactionsByTxIds([
      '6c828920ea3a968f0c3c4a8f14d70b696e0440d8e4e1d019cced1ba2cc63cd51',
      '097cf9d4ec10711e809f316b7738bbbff94efe32ea2cd55e57ddf5840f828741'
    ]);
await client.getOutputsByAddress('13n561iVozTtMXJzAJNA5TQsnTboRvpxae');
await client.getOutputsByAddresses([
      '1P8Jd8qQM7y45iXLM1eiXCCmGRhCPjykZB',
      '16qgC3hzi38xo1vn2gGsNVwWaW1sEH3h9R'
    ]);
await client.getMerkleBranchByTxId('571c7508413415debe4ba146a2ed141e4d4204d0743169ab3366b1f1e1960a5d');
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.0.5](https://gitlab.com/ledgit/xoken-arch-gateway-client/compare/v1.0.4...v1.0.5) (2020-05-14)

# Changelog


src/test/vectors.ts
/**
 * Test vectors generated from python client
 */

export const wire = [
  {
    call: 'getBlockByHeight',
    args: 100,
    hex: '00168400016d4845494748542d3e424c4f434b8182001864',
  },
  {
    call: 'getBlocksByHeights',
    args: [100, 101, 102],
    hex: '001f840001715b4845494748545d2d3e5b424c4f434b5d81820183186418651866',
  },
  {
    call: 'getBlockByHash',
    args: '00000000d1145790a8694403d4063f323d499e655c83426834d4ce2f8dd4a2ee',
    hex:
      '00548400016b484153482d3e424c4f434b818202784030303030303030306431313435373930613836393434303364343036336633323364343939653635356338333432363833346434636532663864643461326565',
  },
  {
    call: 'getBlocksByHashes',
    args: [
      '000000000000000002af2a6de04d4a1a73973827eae348fe4d3f4d05610ff968',
      '000000000000000007fc734cbf1fc04c59cf7ecb6af0707fd5cf5b8d46dc4c75',
    ],
    hex:
      '009b8400016f5b484153485d2d3e5b424c4f434b5d81820382784030303030303030303030303030303030303261663261366465303464346131613733393733383237656165333438666534643366346430353631306666393638784030303030303030303030303030303030303766633733346362663166633034633539636637656362366166303730376664356366356238643436646334633735',
  },
  {
    call: 'getTransactionByTxId',
    args: '7d3eb236b526bd681b7fc499d657d237b4d3bc21ef25b37fc1c70822849f1243',
    hex:
      '005184000168545849442d3e5458818204784037643365623233366235323662643638316237666334393964363537643233376234643362633231656632356233376663316337303832323834396631323433',
  },
  {
    call: 'getTransactionsByTxIds',
    args: [
      '6c828920ea3a968f0c3c4a8f14d70b696e0440d8e4e1d019cced1ba2cc63cd51',
      '097cf9d4ec10711e809f316b7738bbbff94efe32ea2cd55e57ddf5840f828741',
    ],
    hex:
      '00988400016c5b545849445d2d3e5b54585d81820582784036633832383932306561336139363866306333633461386631346437306236393665303434306438653465316430313963636564316261326363363363643531784030393763663964346563313037313165383039663331366237373338626262666639346566653332656132636435356535376464663538343066383238373431',
  },
  {
    call: 'getOutputsByAddress',
    args: '13n561iVozTtMXJzAJNA5TQsnTboRvpxae',
    hex:
      '00398400016e414444522d3e5b4f55545055545d818206782231336e35363169566f7a54744d584a7a414a4e41355451736e54626f527670786165',
  },
  {
    call: 'getOutputsByAddresses',
    args: [
      '1P8Jd8qQM7y45iXLM1eiXCCmGRhCPjykZB',
      '16qgC3hzi38xo1vn2gGsNVwWaW1sEH3h9R',
    ],
    hex:
      '0060840001705b414444525d2d3e5b4f55545055545d8182078278223150384a643871514d3779343569584c4d3165695843436d47526843506a796b5a427822313671674333687a693338786f31766e326747734e56775761573173454833683952',
  },
  {
    call: 'getMerkleBranchByTxId',
    args: '571c7508413415debe4ba146a2ed141e4d4204d0743169ab3366b1f1e1960a5d',
    hex:
      '00568400016d545849442d3e5b4d4e4f44455d818208784035373163373530383431333431356465626534626131343661326564313431653464343230346430373433313639616233333636623166316531393630613564',
  },
];

const _wireByCall = {} as { [key: string]: any };

for (const vector of wire) {
  _wireByCall[vector.call] = vector;
}

export const wireByCall = _wireByCall;


src/test/wire.spec.ts
import 'mocha';
import { wire } from './vectors';
import { Client } from '../lib/client';
import { Socket, createServer } from 'net';
import frame from 'frame-stream';
import { assert } from 'chai';

describe('Wire formats', () => {
  let clientData: Buffer;
  let client: Client;

  before(() => {
    const server = createServer((s: Socket) => {
      // 'connection' listener.
      const inbound = frame.decode({ lengthSize: 2 });
      const outbound = frame.encode({ lengthSize: 2 });

      console.log('client connected');
      s.pipe(inbound);
      outbound.pipe(s);

      s.on('data', (data) => {
        clientData = data;
        outbound.write(Buffer.from('Received something, sending stuff back'));
      });
    });

    server.on('error', (err) => {
      throw err;
    });

    server.listen(19090, () => {
      console.log('Test server bound');
      client = new Client('localhost', 19090);
    });
  });

  for (const vector of wire) {
    it(vector.call, async () => {
      await ((client as unknown) as any)[vector.call](vector.args);
      assert(clientData.equals(Buffer.from(vector.hex, 'hex')));
    });
  }
});


src/cli/cli.ts
/**
 * CLI client work in progress
 */
import { REPLServer } from 'repl';
import commander from 'commander';
import { Client } from '../lib/client';

const program = new commander.Command();
program
  .name('xoken-arch-gateway-client')
  .option('-c, --host', 'Host', '127.0.0.1')
  .option('-p, --port', 'Port', '9090')
  .option('-i, --interactive', 'Interactive mode');

program.command('getBlockByHeight <height>');
program.command('getBlocksByHeights <heights>');
program.command('getBlockByHash <hash>');
program.command('getBlocksByHashes <hashes>');
program.command('getTransactionByTxId <txid>');
program.command('getTransactionsByTxIds <txids>');
program.command('getOutputsByAddress <address>');
program.command('getOutputsByAddresses <addresses>');
program.command('getMerkleBranchByTxId <txid>');

program.parse(process.argv);

if (program.args.length === 0) {
  program.outputHelp();
  process.exit(0);
}

const client = new Client(program.host, parseInt(program.port || 9090));

// eslint-disable-next-line
const repl = require('repl');
const replServer: REPLServer = repl.start('>>> ');

const wrapper = (f: Function) => async (args: any) => {
  const F = f.bind(client);
  F(args).then(console.log);
};

for (const func of Object.getOwnPropertyNames(Client.prototype)) {
  if (func === 'constructor' || func === 'call') continue;
  (replServer.context as any)[func] = wrapper((client as any)[func]);
}
// //const optLength = Object.keys(program.opts()).length;
// const cmd = 'get';
// let args: any = program.args;

// for (let i = 0; i < args.length; i++) {
//   if (!isNaN(parseInt(args[i]))) {
//     args[i] = parseInt(args[i]);
//   }
// }

// if (args.length === 1) {
//   args = args[0];
// }

// (client as any)[cmd](args).then((result: any) => {
//   console.log(JSON.stringify(result, null, 2));
// });


src/lib/client.ts
import frame from 'frame-stream';
import { Socket, createConnection } from 'net';
import cbor from 'cbor';
import { Duplex } from 'stream';

export class Client {
  private socket: Socket;
  private outbound: Duplex;
  private inbound: Duplex;

  constructor(host: string, port: number) {
    this.socket = createConnection(port, host);
    this.outbound = frame.encode({ lengthSize: 2 });
    this.inbound = frame.decode({ lengthSize: 2 });
    this.outbound.pipe(this.socket);
    this.socket.pipe(this.inbound);
  }

  async call(func: string, i: number, args: any): Promise<any> {
    return new Promise((resolve) => {
      const msg = cbor.encode([0, 1, func, [[i, args]]]);
      const handler = (msg: Buffer) => {
        const decoded = cbor.decode(msg);
        this.inbound.off('data', handler);
        resolve(decoded);
      };
      this.inbound.on('data', handler);
      this.outbound.write(msg);
    });
  }

  async getBlockByHeight(height: number): Promise<any> {
    return this.call('HEIGHT->BLOCK', 0, height);
  }

  async getBlocksByHeights(heights: Array<number>): Promise<any> {
    return this.call('[HEIGHT]->[BLOCK]', 1, heights);
  }

  async getBlockByHash(hash: string): Promise<any> {
    return this.call('HASH->BLOCK', 2, hash);
  }

  async getBlocksByHashes(hashes: Array<string>): Promise<any> {
    return this.call('[HASH]->[BLOCK]', 3, hashes);
  }

  async getTransactionByTxId(txid: string): Promise<any> {
    return this.call('TXID->TX', 4, txid);
  }

  async getTransactionsByTxIds(txids: Array<string>): Promise<any> {
    return this.call('[TXID]->[TX]', 5, txids);
  }

  async getOutputsByAddress(address: string): Promise<any> {
    return this.call('ADDR->[OUTPUT]', 6, address);
  }

  async getOutputsByAddresses(addresses: Array<string>): Promise<any> {
    return this.call('[ADDR]->[OUTPUT]', 7, addresses);
  }

  async getMerkleBranchByTxId(txid: string): Promise<any> {
    return this.call('TXID->[MNODE]', 8, txid);
  }
}


contrib/client.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
import sys
import glob
import time
import socket
from cbor2 import dumps, loads, CBORTag


def sendRequest(s, payload):
    length = len(payload)
    lenPrefix = length.to_bytes(2, 'big')
    full = lenPrefix + payload
    s.sendall(lenPrefix)
    print ('Sending lenPrefix', lenPrefix)
    s.sendall(payload)
    print ('Sending payload', payload)
    prefix = s.recv(2)
    leng = int.from_bytes(prefix, byteorder='big')
    raw = s.recv(leng)
    data = loads(raw)
    print ('Received', data)


def client(host, port):
    print ('Starting...')
    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    s.connect((host, int(port)))
    return s


if len(sys.argv) < 3:
    print ('Invalid args, need: <hostname> <port>')
    exit(0)

sock = client(sys.argv[1], sys.argv[2])


x1 = dumps((0, 1, 'HEIGHT->BLOCK', [(0, 100)]))

x2 = dumps((0, 1, '[HEIGHT]->[BLOCK]', [(1, [100, 101, 102])]))

x3 = dumps((0, 1, 'HASH->BLOCK', [(2,
           '00000000d1145790a8694403d4063f323d499e655c83426834d4ce2f8dd4a2ee'
           )]))

x4 = dumps((0, 1, '[HASH]->[BLOCK]', [(3,
           ['000000000000000002af2a6de04d4a1a73973827eae348fe4d3f4d05610ff968'
           ,
           '000000000000000007fc734cbf1fc04c59cf7ecb6af0707fd5cf5b8d46dc4c75'
           ])]))

x5 = dumps((0, 1, 'TXID->TX', [(4,
           '7d3eb236b526bd681b7fc499d657d237b4d3bc21ef25b37fc1c70822849f1243'
           )]))

x6 = dumps((0, 1, '[TXID]->[TX]', [(5,
           ['6c828920ea3a968f0c3c4a8f14d70b696e0440d8e4e1d019cced1ba2cc63cd51'
           ,
           '097cf9d4ec10711e809f316b7738bbbff94efe32ea2cd55e57ddf5840f828741'
           ])]))

x7 = dumps((0, 1, 'ADDR->[OUTPUT]', [(6,
           '13n561iVozTtMXJzAJNA5TQsnTboRvpxae')]))

x8 = dumps((0, 1, '[ADDR]->[OUTPUT]', [(7,
           ['1P8Jd8qQM7y45iXLM1eiXCCmGRhCPjykZB',
           '16qgC3hzi38xo1vn2gGsNVwWaW1sEH3h9R'])]))

x9 = dumps((0, 1, 'TXID->[MNODE]', [(8,
           '571c7508413415debe4ba146a2ed141e4d4204d0743169ab3366b1f1e1960a5d'
           )]))

sendRequest(sock, x1)
sendRequest(sock, x2)
sendRequest(sock, x3)
sendRequest(sock, x4)
sendRequest(sock, x5)
sendRequest(sock, x6)
sendRequest(sock, x7)
sendRequest(sock, x8)
sendRequest(sock, x9)
#
